### ISSUE TITLE
Loading and running trained models

### ISSUE THREAD

**Issue by sharsnik2** [google-research/seed_rl#41]
I've successfully trained a network on my custom environment. Now I'd like to be able to observe the activities of the hidden states as the network navigates the environment. As such, I'm loading the latest checkpoint and trying to mimic the inference/environment step cycle in a single script. I will then gather the agent outputs and analyse them later. Here is the code I'm using is pasted below.

The issue is that while this code produces better than uninitialized returns on the environment, it is not nearly as good as the returns I'm getting in the eval agent during training (or even the non-eval agents). So, it seems that I must be missing something that is done during the training loop. 

One possible thing is that when loading from the checkpoint the leaner wants agent AND target_agent. But, I'm only loading target agent: `ckpt = tf.train.Checkpoint(target_agent=agent)`

```
with strategy.scope():
	@tf.function
	def inference(*args):
		return agent(*decode(args))

observation = env.reset()
reward = 0.0
raw_reward = 0.0
done = False
zeroIndex = tf.constant([0], dtype=tf.int32)

agent_outputs.reset(zeroIndex)
agent_states.replace(zeroIndex, agent.initial_state(1))
		
while not done:
	env_output = utils.EnvOutput(reward, done, observation)
		
	env_outputs.replace(zeroIndex, env_output)
			
	input_ = encode((agent_outputs.read(zeroIndex), env_outputs.read(zeroIndex)))
	agent_output, agent_state = inference(input_, agent_states.read(zeroIndex))
				
	agent_outputs.replace(zeroIndex, agent_output.action)	
	agent_states.replace(zeroIndex, agent_state)	
			
	observation, reward, done, info = env.step(agent_output.action.numpy()[0])
```

**Comment by sharsnik2** [comment]
I have a few updates. I've been running the code with the exact same environment layout each time. As such, my code returns the same total reward each run. The base seed_rl code (on an eval agent with epsilon = 0), however, returns the SAME value as my code on the first run-through, but then different values on all other runs. 

The dynamics of the hidden state are the same for my code and the first run of the seed_rl code but after that, the initial state of the seed_rl code is non-zero, which is part of what causes the divergence. Why do all the repeated runs of the seed_rl eval agent state at a non-zero network state? Is the eval agent also the replay buffer? 

The code above seems to be more or less working, as the first run of the network while training is identical to what I'm getting, but the main issue is that my code has MUCH worse performance than the seed_rl training output would lead me to believe.  Is there a way to get my non-replay buffered runs of the network to mimic the training-time performance?

**Comment by lespeholt** [comment]
Will look into it, but that's indeed counter intuitive.
However, it shouldn't affect the score of the agent really as one can always give the agent an arbitrary reward/action if the 0 action and 0 reward for some reason is a really bad starting point.

**Comment by lespeholt** [comment]
Which agent do you try to load? R2D2 or V-trace?

**Comment by sharsnik2** [comment]
I was loading an r2d2 agent. I also used a custom network based on the atari network.

You can indeed give a non zero action/reward pair as a starting point, but it's a bit tricky to know what range of starting points is likely to be "good". This is even more confusing, as the range of "good" starting points is likely to change as the training takes place (i.e. once the agent starts to win most games the first reward score will increase dramatically).

**Comment by lespeholt** [comment]
I definitely agree. Was just pointing out that it may not be the root cause of why you can't reproduce the training results.

**Comment by sharsnik2** [comment]
Yeah, for sure. My code above does seem to work now that I've cleared the reward and action on environment reset, so I'll close this thread for now. 

Thanks!

**Comment by Edvard-D** [comment]
Hi @sharsnik2, do you possibly still have the full file you used to run a trained model? I have a model of my own I'm trying to run, but am having a hard time understanding how to set everything up correctly. If you'd be willing to share that it would be a huge help!

**Comment by sharsnik** [comment]
Hey @Edvard-D. The code I've been using to run experiments (i.e. observe the activity of a trained agent) is here: https://github.com/sharsnik2/seed_rl/blob/unity/unity/runExperiment.py

I set this up for my custom unity environment and it will probably (definitely) not work out of the box for you, but hopefully it'll be a good starting point. If you have specific questions, I'll be happy to discuss.

**Comment by Edvard-D** [comment]
Thanks for the reply. I did manage to get it working after posting my question, but your link will be very useful for anyone looking to do the same in the future.
